Client/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Paged/CookieControlScript.js

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+window.__firefox__.execute(function($) {
+  //Cookie should neither be saved nor accessed (local or session) when user has blocked all cookies.
+  const cookie = Object.getOwnPropertyDescriptor(Document.prototype, 'cookie') || Object.getOwnPropertyDescriptor(HTMLDocument.prototype, 'cookie')
+  
+  if (cookie && cookie.configurable) {
+    Object.defineProperty(document, 'cookie', {
+      get: $(function() {
+        // Not returning null here as some websites don't have a check for cookie returning null, and may not behave properly
+        return "";
+      }),
+      set: $(function(val) {
+        console.error("Access denied for 'cookie'")
+        return;
+      })
+    });
+  }
+  
+  //Access to localStorage should be denied when user has blocked all Cookies.
+  if (Object.getOwnPropertyDescriptor(window, 'localStorage')) {
+    Object.defineProperty(window, 'localStorage', {
+      get: $(function() {
+        console.error("Access denied for 'localStorage'")
+        return null;
+      }),
+    });
+  }
+  
+  //Access to sessionStorage should be denied when user has blocked all Cookies.
+  if (Object.getOwnPropertyDescriptor(window, 'sessionStorage')) {
+    Object.defineProperty(window, 'sessionStorage', {
+      get: $(function() {
+        console.error("Access denied for 'sessionStorage'")
+        return null;
+      })
+    });
+  }
+  
+  (() => {
+    // Access to caches should be denied when user has blocked all Cookies.
+    const makeFailingPromiseFunction = $(function() {
+      return $(function() {
+        return Promise.reject(
+          new DOMException('An attempt was made to break through the security policy of the user agent.')
+        )
+      });
+    });
+    
+    // We need to check that window.caches is defined as this API was only added in iOS 15
+    // Later on we can probably remove this check.
+    if (window.caches !== undefined) {
+      window.CacheStorage.prototype.match = makeFailingPromiseFunction()
+      window.CacheStorage.prototype.has = makeFailingPromiseFunction()
+      window.CacheStorage.prototype.delete = makeFailingPromiseFunction()
+      window.CacheStorage.prototype.open = makeFailingPromiseFunction()
+      window.CacheStorage.prototype.keys = makeFailingPromiseFunction()
```

Comment:
I was thinking of creating a separate issue for this one, but looks like it would be an easy fix.
So I found an interesting bypass for this. Basically, we overwrite the prototypes here to prevent a website from storing persistent data in Cache when Block all Cookies setting is on. Since we don't freeze the prototypes, they can be restored if an attacker or malicious website could find a way to get the original prototypes of these functions. So I found that I can get prototypes from an iframe and restore it in the main frame. The problem is that this script is injected in all iframes as well, so prototypes would be tainted there as well. 
Eventually, I found that if I load `internal://` URLs in the iframe (e.g `<iframe src="internal://local/errorpage"" id="iframe"></iframe>`), this script is not loaded so that I can access the original prototypes and restore the cache functionality. Worth noting that an internal URL is not loading properly, instead it shows `about:blank` as the origin.

You can test it here https://stoletheminerals.github.io/caches.html. Enable `Block all Cookies` setting in Brave, open the page. Click `Set Tracking ID`, nothing happens. Click `restoreCache` and then `Set Tracking ID` again, you will get an ID. You can then close the browser, open it again, click `restoreCache` and `Get Tracking ID`, you get the same ID.

Calling `Object.freeze(window.CacheStorage.prototype)` should be enough to fix it

cc @cuba @pes10k

Reply:
A follow-up: https://github.com/brave/internal/issues/914